### PR TITLE
Aggies sprint9 staging merge

### PR DESCRIPTION
**Pull request by wwelling**

### CODE REVIEW

**Review by kaladay — CHANGES_REQUESTED** [review]
This needs a setting to allow toggling between using OKAPI and going directly to `mod-workflow` as is done in the previous behavior.

We still need and still use the direct to `mod-workflow` behavior.

File: src/service/okapi.service.ts [hunk 1]
```diff
@@ -32,36 +32,18 @@ class OkapiService extends RestService {
...
-                  foundAccess = matched;
-                } else {
-                  matched = cookies[i].match(matchRefresh);
-
-                  if (!!matched) {
-                    foundRefresh = matched;
-                  }
-                }
+              if (maybe.length > 0) {
+                return maybe[0].substring(token.length + 1, maybe[0].indexOf(';'));
```
Comment (kaladay) [inline src/service/okapi.service.ts:40]: This string `folioAccessToken;` should not match but might (And also `folioAccessTokenBlahBlahBlah=...;`).

Please use more precise matching to at least make sure the format is like that described in the regex: `/folioAccessToken=([^;\s]*)/gi;`, such as `/$token=([^;\s]*)/gi;` (assuming the `$token` variable gets expanded appropriately in a regex statement).

**Review by kaladay — CHANGES_REQUESTED** [review]

**Review by jeremythuff — APPROVED** [review]

File: src/config.ts [hunk 2]
```diff
@@ -32,23 +28,16 @@ const schema = {
...
-    default: 'ad6d42d1-1239-526b-a997-6b96888f7719'
-  },
-  'mod-camunda': {
-    type: 'string',
-    default: 'http://localhost:9000'
-  },
-  'mod-camunda-internal': {
-    type: 'string',
-    default: 'http://localhost:9000'
+    default: 'e4252f55-cc7e-5e1e-a17e-7bfc16706a77'
```
Comment (kaladay) [inline src/config.ts:31]:
```suggestion
    default: ''
```
We shouldn't have a default `userId` in here.

**Review by kaladay — CHANGES_REQUESTED** [review]

File: src/service/okapi.service.ts [hunk 3]
```diff
@@ -30,58 +30,48 @@ class OkapiService extends RestService {
...
+            if (response?.body?.refreshToken) {
               refreshToken = response.body.refreshToken;
-            } else if (!!response.body.folioRefreshToken) {
+            } else if (response.body?.folioRefreshToken) {
               refreshToken = response.body.folioRefreshToken;
             }
           }
 
-          if (!!accessToken) {
+          if (accessToken) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:69]:
```suggestion
          if (!!accessToken) {
```
This is Javascript, we need to use the double negate to enforce boolean.

File: src/service/okapi.service.ts [hunk 4]
```diff
@@ -30,58 +30,48 @@ class OkapiService extends RestService {
...
             config.set('token', accessToken);
             config.set('folioAccessToken', accessToken);
           }
 
-          if (!!refreshToken) {
+          if (refreshToken) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:74]:
```suggestion
          if (!!refreshToken) {
```
This is Javascript, we need to use the double negate to enforce boolean.

File: src/service/okapi.service.ts [hunk 5]
```diff
@@ -30,58 +30,48 @@ class OkapiService extends RestService {
...
+
+            accessToken = extractCookieValue('folioAccessToken');
+            refreshToken = extractCookieValue('folioRefreshToken');
+
           } else {
-            if (!!response?.body?.okapiToken) {
+            if (response?.body?.okapiToken) {
               accessToken = response.body.okapiToken;
-            } else if (!!headers['x-okapi-token']) {
+            } else if (headers?.['x-okapi-token']) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:58]:
```suggestion
            } else if (!!headers?.['x-okapi-token']) {
```
This is Javascript, we need to use the double negate to enforce boolean.

File: src/service/okapi.service.ts [hunk 6]
```diff
@@ -30,58 +30,48 @@ class OkapiService extends RestService {
...
-            refreshToken = Array.isArray(foundRefresh) ? foundRefresh[0] : undefined;
+              throw new Error('Invalid cookie');
+            };
+
+            accessToken = extractCookieValue('folioAccessToken');
+            refreshToken = extractCookieValue('folioRefreshToken');
+
           } else {
-            if (!!response?.body?.okapiToken) {
+            if (response?.body?.okapiToken) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:56]:
```suggestion
            if (response?.body?.okapiToken) {
```
This is Javascript, we need to use the double negate to enforce boolean.

File: src/service/okapi.service.ts [hunk 7]
```diff
@@ -30,58 +30,48 @@ class OkapiService extends RestService {
...
               accessToken = response.body.okapiToken;
-            } else if (!!headers['x-okapi-token']) {
+            } else if (headers?.['x-okapi-token']) {
               accessToken = headers['x-okapi-token'];
             }
 
-            if (!!response?.body?.refreshToken) {
+            if (response?.body?.refreshToken) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:62]:
```suggestion
            if (response?.body?.refreshToken) {
```
This is Javascript, we need to use the double negate to enforce boolean.

File: src/service/okapi.service.ts [hunk 8]
```diff
@@ -30,58 +30,48 @@ class OkapiService extends RestService {
...
               refreshToken = response.body.refreshToken;
-            } else if (!!response.body.folioRefreshToken) {
+            } else if (response.body?.folioRefreshToken) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:64]:
```suggestion
            } else if (response.body?.folioRefreshToken) {
```
This is Javascript, we need to use the double negate to enforce boolean.

File: src/service/okapi.service.ts [hunk 9]
```diff
@@ -30,58 +30,48 @@ class OkapiService extends RestService {
...
-            const matchAccess = /folioAccessToken=([^;\s]*)/gi;
-            const matchRefresh = /folioRefreshToken=([^;\s]*)/gi;
-            const cookies = !!headers['set-cookie'] ? headers['set-cookie'] : undefined;
-            let foundAccess;
-            let foundRefresh;
+          const headers = response?.headers;
 
-            if (!!cookies && Array.isArray(cookies)) {
-              let matched;
+          if (headers?.['set-cookie'] && Array.isArray(headers['set-cookie'])) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:36]:
```suggestion
          if (!!headers?.['set-cookie'] && Array.isArray(headers['set-cookie'])) {
```
This is Javascript, we need to use the double negate to enforce boolean.

**Review by kaladay — CHANGES_REQUESTED** [review]

File: src/service/okapi.service.ts (hunk 9 above)
Comment (wwelling) [reply src/service/okapi.service.ts:33]: Yeah, I thought so as well but it does not pass default linting.

Comment (wwelling) [reply src/service/okapi.service.ts:33]: Oh, that condition is already occurring with `const headers = response?.headers;`

Comment (wwelling) [reply src/service/okapi.service.ts:33]: Nope. It is for headers undefined. Anyway, I will revert to double negation.

Comment (kaladay) [reply src/service/okapi.service.ts]:
```js
var struct = { header: "header" };
var header = struct?.header;

console.log(typeof header);
console.log(typeof !!header);
```

Produces:
```
 "string"
 "boolean"
```

We want to ensure that the comparison is a boolean comparison.

File: src/service/okapi.service.ts [hunk 10]
```diff
@@ -30,48 +25,38 @@ class OkapiService extends RestService {
...
               accessToken = response.body.okapiToken;
-            } else if (!!headers['x-okapi-token']) {
+            } else if (!!headers?.['x-okapi-token']) {
               accessToken = headers['x-okapi-token'];
             }
 
             if (!!response?.body?.refreshToken) {
               refreshToken = response.body.refreshToken;
-            } else if (!!response.body.folioRefreshToken) {
+            } else if (!!response.body?.folioRefreshToken) {
```
Comment (kaladay) [inline src/service/okapi.service.ts:59]:
```suggestion
            } else if (!!response?.body?.folioRefreshToken) {
```
I just noticed the missing `?` before body.
It was missing in the original code too.

**Review by kaladay — CHANGES_REQUESTED** [review]

**Review by kaladay — APPROVED** [review]

**Review by jeremythuff — APPROVED** [review]

**Pull request merged**